source/beamline/BioXAS/BioXASCarbonFilterFarmFilterControl.cpp

```diff
@@ -0,0 +1,308 @@
+#include "BioXASCarbonFilterFarmFilterControl.h"
+#include "actions3/AMActionSupport.h"
+#include "actions3/AMListAction3.h"
+
+BioXASCarbonFilterFarmFilterControl::BioXASCarbonFilterFarmFilterControl(const QString &name, const QString &units, QObject *parent) :
+	AMPseudoMotorControl(name, units, parent)
+{
+	// Initialize inherited variables.
+
+	value_ = -1;
+	setpoint_ = -1;
+	minimumValue_ = 0;
+	maximumValue_ = 0;
+
+	setContextKnownDescription("Filter Control");
+	setAllowsMovesWhileMoving(false);
+
+	// Initialize local variables.
+
+	upstreamFilter_ = 0;
+	downstreamFilter_ = 0;
+}
+
+BioXASCarbonFilterFarmFilterControl::~BioXASCarbonFilterFarmFilterControl()
+{
+
+}
+
+bool BioXASCarbonFilterFarmFilterControl::canMeasure() const
+{
+	bool result = false;
+
+	if (isConnected())
+		result = ( upstreamFilter_->canMeasure() && downstreamFilter_->canMeasure() );
+
+	return result;
+}
+
+bool BioXASCarbonFilterFarmFilterControl::canMove() const
+{
+	bool result = false;
+
+	if (isConnected())
+		result = ( upstreamFilter_->canMove() && downstreamFilter_->canMove() );
+
+	return result;
+}
+
+bool BioXASCarbonFilterFarmFilterControl::canStop() const
+{
+	bool result = false;
+
+	if (isConnected())
+		result = ( upstreamFilter_->canStop() && downstreamFilter_->canStop() );
+
+	return result;
+}
+
+bool BioXASCarbonFilterFarmFilterControl::validValue(double value) const
+{
+	bool result = (value >= 0 && value < enumNames().count());
+	return result;
+}
+
+bool BioXASCarbonFilterFarmFilterControl::validSetpoint(double value) const
+{
+	bool result = (value >= 0 && value < moveEnumNames().count());
+	return result;
+}
+
+double BioXASCarbonFilterFarmFilterControl::filterAt(int index) const
+{
+	double result = -1;
+
+	if (index >= 0 && index < filters_.count())
+		result = filters_.at(index);
+
+	return result;
+}
+
+int BioXASCarbonFilterFarmFilterControl::indexOf(double filter) const
+{
+	return filters_.indexOf(filter);
+}
+
+int BioXASCarbonFilterFarmFilterControl::indexOf(const QString filterString) const
+{
+	return enumNames().indexOf(filterString);
+}
+
+void BioXASCarbonFilterFarmFilterControl::setUpstreamFilter(BioXASCarbonFilterFarmActuatorFilterControl *newControl)
+{
+	if (upstreamFilter_ != newControl) {
+
+		if (upstreamFilter_)
+			removeChildControl(upstreamFilter_);
```

Comment:
I think that if you are adding signals at a certain level, that level should also do the disconnecting.  I think it helps with consistency.